src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/FormattingPassBase.cs

```diff
@@ -224,81 +317,128 @@ protected static SourceText CleanupDocument(FormattingContext context, Range ran
+            //
+            //              |}
+            // }
+            // We want to return the length of the range marked by |...|
+            //
             var text = context.SourceText;
-            range ??= TextSpan.FromBounds(0, text.Length).AsRange(text);
-            var csharpDocument = context.CodeDocument.GetCSharpDocument();
+            var sourceMappingSpan = sourceMappingRange.AsTextSpan(text);
+            var whitespaceLength = text.GetFirstNonWhitespaceOffset(sourceMappingSpan);
+            if (whitespaceLength == null)
+            {
+                // There was no content here. Skip.
+                return;
+            }
 
-            var changes = new List<TextChange>();
-            foreach (var mapping in csharpDocument.SourceMappings)
+            var spanToReplace = new TextSpan(sourceMappingSpan.Start, whitespaceLength.Value);
+            if (!context.TryGetIndentationLevel(spanToReplace.End, out var contentIndentLevel))
             {
-                var mappingSpan = new TextSpan(mapping.OriginalSpan.AbsoluteIndex, mapping.OriginalSpan.Length);
-                var mappingRange = mappingSpan.AsRange(text);
-                if (!range.LineOverlapsWith(mappingRange))
-                {
-                    // We don't care about this range. It didn't change.
-                    continue;
-                }
+                // Can't find the correct indentation for this content. Leave it alone.
+                return;
+            }
 
-                var mappingStartLineIndex = mappingRange.Start.Line;
-                if (context.Indentations[mappingStartLineIndex].StartsInCSharpContext)
-                {
-                    // Doesn't need cleaning up.
-                    // For corner cases like (Range marked with |...|),
-                    // @{
-                    //     if (true} { <div></div>| }|
-                    // }
-                    // We want to leave it alone because tackling it here is really complicated.
-                    continue;
-                }
+            // At this point, `contentIndentLevel` should contain the correct indentation level for `}` in the above example.
+            var replacement = context.NewLineString + context.GetIndentationLevelString(contentIndentLevel);
 
-                if (!ShouldCleanup(context, mappingRange.Start))
-                {
-                    // We don't want to run cleanup on this range.
-                    continue;
-                }
+            // After the below change the above example should look like,
+            // @{
+            //     if (true)
+            //     {
+            //         <div></div>
+            //     }
+            // }
+            var change = new TextChange(spanToReplace, replacement);
+            changes.Add(change);
+        }
 
+        private void CleanupSourceMappingEnd(FormattingContext context, Range sourceMappingRange, List<TextChange> changes)
```

Comment:
Wonder if "Cleanup" is the right verbiage. This method is all about separating language contexts. Also, it adds to the changes collection

Reply:
I agree but all the other names I can think of are too long. I'll try to think of a better name.

Reply:
I tried really hard but couldn't come up with a concise name :(